### PR TITLE
Support string translation around the solver boundry

### PR DESCRIPTION
**Pull request by KinanBab**
* String values are read from the p4rt table entries
* They are translated to bit-vectors of appropriate size prior to using them in constraints or the solver.
* Mapping between strings and bit-vectors are maintained for consistent translation and reverse translation.
* Outputs acquired from the constraints are reverse translated, and shown as strings in the output of this tool.
* Stored bit-vectors are normalized to avoid issues reverse translating them in cases of padding or formatting transformations.
* All added code is either not executed or is executed without effects (no-op) for fields and values that do not need translation.

### CODE REVIEW

File: p4_symbolic/symbolic/values.cc [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+#include "p4_symbolic/symbolic/operators.h"
+#include "p4_symbolic/symbolic/symbolic.h"
+
+namespace p4_symbolic {
+namespace symbolic {
+namespace values {
+
+namespace {
+
+// Finds the minimum bit size required for representing the given value.
```
Comment (KinanBab) [inline p4_symbolic/symbolic/values.cc:43]: Below is new code

File: p4_symbolic/symbolic/values.cc [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    if (stripped.at(i) == '1') {
+      first_zero_index = i;
+      break;
+    }
+  }
+  return stripped.substr(first_zero_index);
+}
+
+}  // namespace
+
```
Comment (KinanBab) [inline p4_symbolic/symbolic/values.cc:96]: End of new code, below is a portion of existing code that is copied from a different file for organization purposes.

File: p4_symbolic/symbolic/values.cc [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+      ASSIGN_OR_RETURN(hi, operators::Times(hi, shift));  // shift << 64.
+      return operators::Plus(hi, lo);
+    }
+
+    default:
+      return absl::UnimplementedError(
+          absl::StrCat("Found unsupported value type ", value.DebugString()));
+  }
+}
+
```
Comment (KinanBab) [inline p4_symbolic/symbolic/values.cc:197]: New code below here and until the end of file.

File: p4_symbolic/symbolic/util.cc [hunk 4]
```diff
@@ -172,113 +173,6 @@ gutil::StatusOr<SymbolicTrace> MergeTracesOnCondition(
   return merged;
 }
```
Comment (KinanBab) [inline p4_symbolic/symbolic/util.cc:174]: This code is not really removed, it is moved to another file (values.cc) for organization purposes. I mark its location there with a comment to simplify review.

File: p4_symbolic/symbolic/util.h [hunk 5]
```diff
@@ -53,11 +54,49 @@ gutil::StatusOr<SymbolicTrace> MergeTracesOnCondition(
...
+// The mapping of strings to bitvectors is stored and used to map the same
+// string to the same consistent bitvector, and to also do a reverse-translation
+// when extracting values for headers and fields from the z3 model.
+gutil::StatusOr<z3::expr> P4RTValueZ3Expr(const std::string field_name,
+                                          const pdpi::IrValue &value);
+
+// Exposes a static mapping from string values to bitvector values.
+// We do not need to include field names here because we assume the string
+// values are unique.
+std::unordered_map<std::string, uint64_t> &StringToBitVectorTranslationMap();
```
Comment (konne88) [inline p4_symbolic/symbolic/util.h]: It's unusual that a mutable global variable is the right way to go. What's the reason for not passing around the mapping to make the dataflow more explicit?

**Review by konne88 — CHANGES_REQUESTED** [review]

File: p4_symbolic/symbolic/util.h (hunk 5 above)
Comment (KinanBab) [reply p4_symbolic/symbolic/util.h]: It kind of reminds me of the z3_context and how that one is global too. I think passing it around is fine. Probably the best way to go about it is to have this be a class, with these maps being private in it, and having the parsing/translation functions be part of it, and then this could be passed around as part of the SymbolicPerPacketState or next to it as a new parameter.

Comment (KinanBab) [reply p4_symbolic/symbolic/util.h]: I did not want to pass translation maps around explicitly through out the symbolic evaluation pipeline because they would be too distracting, they are only used in like two places. But I think organizing them as some class/struct, might make that concern mute.

File: p4_symbolic/symbolic/action.cc [hunk 6]
```diff
@@ -326,7 +326,8 @@ absl::Status EvaluateAction(const ir::Action &action,
         parameters.at(i);
     const std::string &parameter_name = parameter.param().name();
     ASSIGN_OR_RETURN(z3::expr parameter_value,
-                     util::IrValueToZ3Expr(args.at(i - 1).value()));
+                     environment->value_formatter.FormatP4RTValue(
+                         parameter_name, args.at(i - 1).value()));
```
Comment (konne88) [inline p4_symbolic/symbolic/action.cc]: Parameter name is the wrong thing there, I think it should be parameter type (vrf_t), right? Like, VRFs in an L3 match should be translated consistently with the action that sets it in the VRF classifier table?

File: p4_symbolic/symbolic/values.h [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+// as a logical translation at the boundry between P4RT and bmv2.
+gutil::StatusOr<z3::expr> FormatBmv2Value(const pdpi::IrValue &value);
+
+// This class is responsible for formatting various values into consistent
+// format acceptable by z3.
+// The class transforms pdpi::IrValue instances to z3::expr, it also handles
+// translation (and reverse translation) of P4RT string values that are meant
+// to be translated to bitvectors, because of @p4runtime_translation
+// annotations.
+class ValueFormatter {
```
Comment (konne88) [inline p4_symbolic/symbolic/values.h]: This class does too many things for my taste. The reason that you want a class, is to maintain the invariant of an id allocator. But this class also does all sort of other things (it does formatting, is a bidirectional map for many fields). Can we split it out.

I would create a new class:

class IdAllocator {
  // If an id has already been allocated, return the same value.
  int32 AllocateId(string);
  StatusOr<string> IdToString(int32) const;
}

typedef P4RuntimeTranslationIdAllocators = std::unordered_map<std::string, IdAllocator>

File: p4_symbolic/symbolic/packet.cc [hunk 8]
```diff
@@ -56,20 +55,59 @@ SymbolicPacket ExtractSymbolicPacket(SymbolicPerPacketState state) {
...
 }
 
-ConcretePacket ExtractConcretePacket(SymbolicPacket packet, z3::model model) {
-  return {model.eval(packet.eth_src, true).to_string(),
-          model.eval(packet.eth_dst, true).to_string(),
-          model.eval(packet.eth_type, true).to_string(),
+gutil::StatusOr<ConcretePacket> ExtractConcretePacket(
+    SymbolicPacket packet, z3::model model,
+    const values::ValueFormatter &value_formatter) {
+  ASSIGN_OR_RETURN(
```
Comment (konne88) [inline p4_symbolic/symbolic/packet.cc]: Why does any of this change? IP addresses etc don't need translation. In fact, nothing in the packet ever needs translation.

File: p4_symbolic/symbolic/packet.h [hunk 9]
```diff
@@ -30,7 +32,9 @@ SymbolicPacket ExtractSymbolicPacket(SymbolicPerPacketState state);
 
 // Extract a concrete packet by evaluating every field's corresponding
 // expression in the model.
-ConcretePacket ExtractConcretePacket(SymbolicPacket packet, z3::model model);
+gutil::StatusOr<ConcretePacket> ExtractConcretePacket(
+    SymbolicPacket packet, z3::model model,
+    const values::ValueFormatter &value_formatter);
```
Comment (konne88) [inline p4_symbolic/symbolic/packet.h]: Packets should never need translation. After-all, the hardware can't do translation either.

File: p4_symbolic/symbolic/symbolic.h [hunk 10]
```diff
@@ -44,6 +45,12 @@ namespace symbolic {
 // evaluation.
 z3::context &Z3Context();
 
+// This struct contains any metadata or information that needs to be passed
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.h]: Why create a struct with just one entry? I think I'd prefer not to have the indirection, so when I read the code, I don't have to jump to this definition to know what's actually going on. If this is in anticipation of having more fields in the future, let's do it in the future when the usecase arrises.

File: p4_symbolic/symbolic/action.cc (hunk 6 above)
Comment (KinanBab) [reply p4_symbolic/symbolic/action.cc]: This is a very important point, we cannot do that currently unfortunately, primarily because we do not have access to the annotation on the type, it is part of p4info, and I could not find it exposed by p4-pdpi, but most important becasue the scenario that you are describing would not use the name "vrf_t" in the bmv2 json (and thus in the action implementation for example), because the bmv2 just inlines the type def directly.

In my docs, I discuss this at length, the current approach will work for the sai program, I discuss why there, but may not work for other programs if they are too liberal in how they use their string values because of what you are hinting at.

Do you have some time to discuss this today?

File: p4_symbolic/symbolic/packet.cc (hunk 8 above)
Comment (KinanBab) [reply p4_symbolic/symbolic/packet.cc]: I can remove this. However, if in the future, some translation happens there, the tool will break. If no translation is needed, the function call is the identity.

File: p4_symbolic/symbolic/symbolic.h (hunk 10 above)
Comment (KinanBab) [reply p4_symbolic/symbolic/symbolic.h]: ok makes sense.

File: p4_symbolic/symbolic/action.cc (hunk 6 above)
Comment (stefanheule) [reply p4_symbolic/symbolic/action.cc]: I may not have all of the context, but PDPI should give you all the info you need. A parameter that is string encoded would look like this:

```
        params_by_name {
          key: "str"
          value {
            param {
              id: 5
              name: "str"
              type_name {
                name: "string_id_t"
              }
            }
            format: STRING
          }
        }
```

You know that this has the type string_id_t, and you know it has the @p4rt_translation(.., string) annotation, because format == Format::STRING.

Is that not enough?

Comment (KinanBab) [reply p4_symbolic/symbolic/action.cc]: I am using this now, it is enough for this. I phrased what I am saying badly.

Everything related to table entries or p4info is sufficiently handled by pdpi. The problem is with things that do not appear in p4info and are only defined in bmv2 JSON (e.g. a header field). All type information about these things is lost because bmv2 strips it out.

Luckily, we can do something that works most of the time. A field that is translated is probably gonna be used as some table match key somewhere (otherwise why is it used), if thats the case, we can find out the type information from p4info/pdpi during the match key evaluation, and store it, so we are now aware of the type of the field. Not 100% fool proof but works well for most cases (and for the SAI program).

That's how the current implementation works. I updated the docs as well.
